gpu_stereo_image_proc_common/src/msg_conversions.cpp

```diff
@@ -2,10 +2,91 @@
 
 #include <ros/ros.h>
 #include <sensor_msgs/image_encodings.h>
+#include <cv_bridge/cv_bridge.h>
 
 using namespace sensor_msgs;
 using namespace stereo_msgs;
 
+DisparityImageGenerator::DisparityImageGenerator (
+    const ImageConstPtr &image, const cv::Mat_<int16_t> disparity16,
+    const image_geometry::StereoCameraModel &model, int min_disparity,
+    int max_disparity, int border, float downsample) {
+
+  DisparityImagePtr disp_msg = boost::make_shared<DisparityImage>();
+  disp_msg->header = image->header;
+  disp_msg->image.header = image->header;
+
+  const int DPP = 16;               // disparities per pixel
+  const double inv_dpp = 1.0 / DPP; // downsample / DPP
+
+  // Fill in DisparityImage image data, converting to 32-bit float
+  sensor_msgs::Image &dimage = disp_msg->image;
+  dimage.height = disparity16.rows;
+  dimage.width = disparity16.cols;
+  dimage.encoding = sensor_msgs::image_encodings::TYPE_32FC1;
+  dimage.step = dimage.width * sizeof(float);
+  dimage.data.resize(dimage.step * dimage.height);
+  cv::Mat_<float> dmat(dimage.height, dimage.width, (float *)&dimage.data[0],
+                       dimage.step);
+
+  // We convert from fixed-point to float disparity and also adjust for any
+  // x-offset between the principal points: d = d_fp*inv_dpp - (cx_l - cx_r)
+  disparity16.convertTo(dmat, dmat.type(), inv_dpp,
+                        -(model.left().cx() - model.right().cx()) /
+                            downsample);
+  ROS_ASSERT(dmat.data == &dimage.data[0]);
+  /// @todo is_bigendian? :)
+
+  const int left = max_disparity + border - 1;
+  const int wtf = (min_disparity >= 0) ? border + min_disparity
+                                       : std::max(border, -min_disparity);
+  const int right = disp_msg->image.width - 1 - wtf;
+  const int top = border;
+  const int bottom = disp_msg->image.height - 1 - border;
+  cv::Rect valid_window(left, top, right - left, bottom - top);
+
+  disp_msg->valid_window.x_offset = valid_window.x;
+  disp_msg->valid_window.y_offset = valid_window.y;
+  disp_msg->valid_window.width = valid_window.width;
+  disp_msg->valid_window.height = valid_window.height;
+
+  // Stereo parameters
+  disp_msg->f = model.right().fx() / downsample;
+  disp_msg->T = model.baseline();
+
+  /// @todo Window of (potentially) valid disparities
+
+  // Disparity search range
+  disp_msg->min_disparity = min_disparity;
+  disp_msg->max_disparity = max_disparity;
+  disp_msg->delta_d = inv_dpp;
+
+  // Create the depth image
+  ImagePtr depth_msg = boost::make_shared<Image>();
```

Comment:
Since we always use the disparity image, but don't always use the depth image (like for the debug disparity images), the Depth image creation should be just-in-time in the getDepth()  function.   

Save whatever state you need to do that calculation ... I think you have most if not all of it in the disparity message.

FWIW I don't think you need to store the Depth image but I don't think it makes a huge difference.

Reply:
This is a good optimization - I can easily save the intermediate Mat and move the processing to getDepth(). I'll focus on making sure the depth image is correct before doing this.